src/M2ulPhyS.cpp

```diff
@@ -299,6 +351,21 @@ void M2ulPhyS::initVariables() {
   }
   assert(mixture != NULL);
 
+#if defined(_CUDA_)
+  GasMixture **d_mixture_tmp;
+  cudaMalloc((void **)&d_mixture_tmp, sizeof(GasMixture **));
+  instantiateDeviceMixture<<<1, 1>>>(config.workFluid, config.GetEquationSystem(), config.visc_mult, config.bulk_visc,
+                                     dim, nvel, d_mixture_tmp);
+  cudaMemcpy(&d_mixture, d_mixture_tmp, sizeof(GasMixture *), cudaMemcpyDeviceToHost);
```

Comment:
For cuda, the memory is allocated for a size of `GasMixture`, while for hip it is allocated for `DryAir`. Is this an intended difference? If so, is hip required to allocate precisely for the inherited class that will be used?

Reply:
First, this difference was intentional, but it could (and probably should) be removed.  The difference is due to the fact that I did not want to use device `malloc` or `new` in hip (because device `new` isn't supported and device `malloc` still seems experimental in that one has to use `-D__HIP_ENABLE_DEVICE_MALLOC__`).  I'd already coded the cuda version using device `new`, but I think it could be refactored to be exactly analogous the hip version.

Second, in the hip version, space for `DryAir` is allocated because that is what is needed before constructing the `DryAir` object via the placement `new` call on the device.  The cuda version allocates that memory via the `new` call on the device.  The `cudaMalloc` call you're referring to here is allocating space for a pointer to `GasMixture`.

Finally, regarding 

> If so, is hip required to allocate precisely for the inherited class that will be used?

the answer is yes, but so is the cuda version.  It just does it by calling `new` with `DryAir`.  This is analogous to the cpu variant also, where the logic in the preceeding `switch` determines which objects to construct.  This kind of logic will eventually have to appear in the device version, but since only `DryAir` is currently supported, it just constructs `DryAir` for now.